Add Transformer API Interface and `@cirq.transformer` decorator

tanujkhattar opened the pull request:
Defines `TRANSFORMER_TYPE` and Implements the `@transformer` decorator, as proposed in https://tinyurl.com/cirq-circuit-transformers-api

All existing transformers will be rewritten to follow the new API once this is checked-in. 
Implementation of `TransformerStatsLoggerBase` will follow in a separate PR. 

Part of #4483

cc @maffoo PTAL at all the mypy magic.

MichaelBroughton commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,234 @@
+    overload,
+)
+import dataclasses
+from enum import Enum
```
Nit: module import.

tanujkhattar replied:
done

MichaelBroughton commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,234 @@
+class TransformerContext:
+    """Stores common configurable options for transformers."""
+
+    logger: Optional[TransformerStatsLoggerBase] = None
```
Would we want to allow this to be optional ? Wondering if it would be better for things implementing this to always log at different levels and then leave it up to the user to set the level of reporting when they call a particular transformer.

tanujkhattar replied:
Since the logger is stateful, a user code which needs to use the logging when calling multiple optimization stages would need to explicitly pass a user generated `context` to each stage. i.e.

```python
>>> context = TransformerContext(logger)
>>> circuit = transformer_stage1(circuit, context)
>>> circuit = transformer_stage2(circuit, context)
...
```
whereas if they are not interested in the logs, they could simply do 
```python
>>> circuit = transformer_stage1(circuit)
>>> circuit = transformer_stage2(circuit)
```

assuming each transformer stage has a default `context=TransformerContext()` specified. If we make the logger as a required field, we'd expect users to always write way-1. I'm not sure if we should necessarily impose this restriction on users. LMK if you feel otherwise.

MichaelBroughton replied:
Ahhh good point. Let's keep it optional.

MichaelBroughton commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,234 @@
+    WARNING = 3
+
+
+class TransformerStatsLoggerBase:
```
Why do we need a Base class + Implementation class here. It seems like the registerinitial and registerfinal would be the same for all transformers and the only thing changing would the circuits we are registering and the names. Am I missing something ?

tanujkhattar replied:
The different implementations would enable different kinds of loggers. For example, 
- One implementation could be a logger which works well for jupyter notebooks / html environments and plots an interactive toolbar to scroll through the action of different stages.
- While another logger could be more suited for a terminal based environment which outputs nicely formatted text to stdout / file. 

Different loggers could also choose to extract different kinds of information from the initial and final circuits passed for each optimization stage (eg: delta of number of 2q gates, no. of microwave gates etc.)

MichaelBroughton replied:
That makes sense. I think I'm sold on doing this, one thing that still worries me a little is the erosion of precision on the different logging levels as the number of classes grow. So what does INFO vs WARNING mean in terminal vs on jupyter vs someone defining it for themselves in their own subclass implementation. Why couldn't one logger class gather all this data and then we leave it to the user to pull out no. of microwave gates of circuit depth or whatever ?

tanujkhattar replied:
- There are two places where the precision levels will be used -- one at the time of logging inside a transformer and second at the time of viewing the logs by the user. The first part will be the same for all logger implementations while the second part will be different for each logger. So, to answer your question, the logs for INFO or WARNING received by each logger implementation would be the same. But how would these different precision-level logs get rendered on a terminal vs a jupyter notebook would vary based on the subclass implementation. 
- One logger could gather all the data -- i.e `(initial_circuit, final_circuit, <list of all intermediate logs>)` for each transformer stage and store it. We could then make abstract methods for `logger.plot` / `logger.show_logs` which the subclasses could implement and do whatever data processing is needed at a later stage. This approach has a drawback that it limits us from writing more efficient loggers which need not necessarily store all this information. 

We could also provide the default implementation in the base class which stores all the data and let users who care about efficiency override these methods to define their custom loggers. I don't have strong opinions either way, but I'd prefer to do the implementation in a separate PR to keep the focus on the API implementation here.

MichaelBroughton replied:
> We could also provide the default implementation in the base class which stores all the data and let users who care about efficiency override these methods to define their custom loggers

Let's do that and see how far we get :)

tanujkhattar replied:
You want me to do that in the current PR itself? I can send a follow-up PR since the two changes are unrelated (the transformer API vs logger implementation).

tanujkhattar replied:
Renamed the base class to `TransformerStatsLogger`.

MichaelBroughton commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,234 @@
+    """
+    if isinstance(cls_or_func, type):
+        return transformer_class(cls_or_func)
+    else:
```
Nit don't need the else block here.

tanujkhattar replied:
done

MichaelBroughton commented on cirq-core/cirq/transformers/transformer_api.py:
````diff
@@ -0,0 +1,234 @@
+    The decorated class must implement the `__call__` method satisfying the `TRANSFORMER_TYPE` API.
+
+    For example:
+    ```
+    @transformer_class
+    class ConvertToSqrtISwaps:
+        def __init__(self):
+            ...
+        def __call__(circuit: cirq.Circuit, context: cirq.TransformerContext) -> cirq.Circuit:
+            ...
+    ```
````
Code blocks should use
```
>>> class Thing
...    Indented thing in class

>>> test(1,2,3)
<output>
```

tanujkhattar replied:
done

MichaelBroughton reviewed (commented):
This looks good. One high level question is do we need an interface for the statslogger or could we just make it concrete and put the implementation for register_initial and final inside that class ? I'm having trouble seeing what major differences each subclass might have.

tanujkhattar reviewed (commented):
Addressed the comments. 

For the logger, each subclass can chose to process, store and plot different information for the registered initial/final circuits (and their delta).

maffoo commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,229 @@
+
+
+def transformer_class(cls: TRANSFORMER_CLS_TYPE) -> TRANSFORMER_CLS_TYPE:
+    """Decorator to verify API and append logging functionality to callable transformer class.
```
Instead of a decorator, can we do this with a base class that has an `__init_subclass__` method? Something like this:
```python
class Transformer(abs.ABC):
    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)
        # Note: would have to modify `transformer_method` to accept an optional `name` to use instead of `func.__name__`
        cls.__call__ = transformer_method(cls.__call__, name=cls.__name__)

class ConvertToSqrtISwaps(Transformer):
    def __call__(self, circuit.Circuit, context: cirq.TransformerContext) -> cirq.Circuit:
        ...
```
I find that class decorators can be rather brittle and `__init_subclass__` handles most cases.

maffoo commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,229 @@
+    pass
+
+
+def transformer(
```
I'm not sure having a combined API for methods is all that helpful versus having a function decorator and a base class:
```python
@cirq.transformer
def my_func(circuit, context):
    ...

class MyCls(cirq.Transformer):
    def __call__(self, circuit, context):
        ...
```

tanujkhattar replied:
My intuition was that having a combined API can be easier to signify that we are doing the same thing in both the cases, i.e adding logging functionality to the transformer. Having two separate ways for classes and methods can potentially be confusing. 

But I don't have strong opinions here and would be happy to change if you think this is a common pattern. This would certainly make dealing with the mypy types a lot simpler.

maffoo replied:
I guess I'm ok with the combined `cirq.transformer` call.

maffoo commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,229 @@
+)
+
+
+def _transformer_with_logging(
```
nit: I'd suggest naming this with a verb phrase, since it actually _calls_ the transformer (name sounds like it produces a transformer with logging added). Perhaps `_transform_and_log`?

tanujkhattar replied:
done

tanujkhattar commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,229 @@
+    max_parallelism: int = 1
+
+
+TRANSFORMER_TYPE = Callable[[CIRCUIT_TYPE, TransformerContext], CIRCUIT_TYPE]
```
Since argument to a `Callable` are contravariant, defining the transformer type to be `Callable[[cirq.AbstractCircuit, TransformerContext], cirq.AbstractCircuit]` would mean that a `Callable[[cirq.Circuit, TransformerContext], cirq.Circuit]]` would not be accepted as a valid transformer. 
So, every transformer would need to accept `cirq.AbstractCircuit` and deal with all subtypes (currently, this just means calling `circuit = abstract_circuit.unfreeze()`, which can probably go in the decorator itself to avoid redundancy).  However, a much bigger issue is that every possible transformer would need to deal with any future subtypes of `AbstractCircuit` that we might end up adding, which is bad.


If we enforce that a transformer should preserve the input type, then we move the responsibility of passing in the correct type on to the user.  Though I agree that it can be annoying for the user to do the freeze/unfreeze conversions manually and we should probably consider adding it to the decorator somehow. 

What do you think?

maffoo replied:
It's hard to decide on the best design here without seeing how people are going to use transformers, in particular how they will be composed. The main advantage I can see to carefully preserving the specific circuit types in these type signatures is that you can have the type checker tell you when calling one or another transformer that you're passing the right circuit types around and haven't forgotten a `freeze` or `unfreeze` call somewhere. But that only works if you are defining the composed transformer statically and running mypy on the code. For people composing transformers at runtime these precise types and associated type gymnastics won't help, and we'd probably want to insert runtime type checks and do the `freeze` and `unfreeze` calls to make it ergonomic for people to compose transformers (or even just to use them, for that matter). But in that case the types don't really help because it's hard for the code doing the composition to see what types were declared by the transformers it is composing (runtime reflection of mypy types is hard and not particularly well supported).

I think I'd lean toward doing the simplest thing possible by accepting and returning `cirq.AbstractCircuit`, which simplifies the API here and makes it the responsibility of the transformer implementer to call `freeze`/`unfreeze` as needed.

If we do stick with the more specific types, I'd still suggest calling this `TRANSFORMER` instead of `TRANSFORMER_TYPE`.

tanujkhattar replied:
How about this?

```python
TRANSFORMER = Callable[[cirq.Circuit, cirq.TransformerContext], cirq.Circuit]
TRANSFORMER_OUT = Callable[[cirq.AbstractCircuit, cirq.TransformerContext], cirq.Circuit]

def transformer(cls_or_func: TRANSFORMER) -> TRANSFORMER_OUT:
    ...
```

- Since the user is calling a "transformer" on an input circuit, it's reasonable to assume that the returned circuit type would always be a mutable circuit (similar to how calling `sorted(input_tuple)` returns a `list`). 
- The `@cirq.transformer` decorator will automatically pass `input_circuit.unfreeze(copy=True)` to the decorated transformer as input. This serves two purposes -- we avoid the redundant `unfreeze()` in each transformer and ensure that the transformer always receives a copy of the input circuit, thus ensuring that the input is never mutated (which is also supposed to be part of the new API contract). 
- A `Callable[[cirq.Circuit, TransformerContext], cirq.Circuit]]` would now be a valid transformer when type checked using the `@transformer` decorator. 
- Since the returned type is `Callable[[cirq.AbstractCircuit, cirq.TransformerContext], cirq.Circuit]`, circuit composition would now work for both frozen and mutable circuits. 


What do you think?

tanujkhattar commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,229 @@
+
+    INFO = 1
+    VERBOSE = 2
+    WARNING = 3
```
I'm not sure if we need such as wide range as: TRACE, DEBUG, INFO, WARN, ERROR, FATAL. 

I'll add explanations and reorder as:
```
WARNING = 1
INFO = 2
VERBOSE = 3
```

tanujkhattar replied:
done

tanujkhattar commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,229 @@
+
+@dataclasses.dataclass()
+class TransformerContext:
+    """Stores common configurable options for transformers."""
```
Any instance of a `Hashable` type is a valid tag. Added explanation of the attributes for clarity.

tanujkhattar commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,229 @@
+
+    logger: Optional[TransformerStatsLoggerBase] = None
+    ignore_tags: Tuple[Hashable, ...] = ()
+    max_parallelism: int = 1
```
Removed `max_parallelism` for now.

tanujkhattar commented:
@maffoo @dstrain115 I've resolved most of the comments. The two big pending questions are:

- What should be the mypy type of the TRANSFORMER? See the discussion thread below for my comments. 
- Do you have strong opinions of whether we should we use two different ways -- base class and decorators, for appending the same logging functionality to class vs method transformers? (see comment on the discussion thread below)

tanujkhattar commented:
@MichaelBroughton @dstrain115 @maffoo I've now also added the implementation of `TransformerStatsLogger` in this PR. Some additional changes are:
- `logger: TransformerStatsLogger` is not an optional field anymore and defaults to `NoOpTransformerStatsLogger()`, which ignores all logging calls to avoid any performance degradations. This is done mainly to avoid the inconvenience of checking `if context.logger is not None` every time a user implementing a transformer wants to log something. 
- `TransformerStatsLogger` implements a logger which stores the initial/final circuits and the corresponding logs added in between for an active transformer. The implementation maintains an implicit dfs ordered graph for the nested transformer cases and a call to the `show()` method prints out the DFS ordered logs of each transformer stage (nested stages are appropriately indented). 
- Taking inspiration from common logging frameworks, I've renamed `VERBOSE` to `DEBUG` in `LogLevel`s. 

One TODO left is to add support for writing the logs to a file instead of printing to stdout when `logger.show()` is called (the feature would be a controlled by an additional parameter). I'll add this in a separate PR. 

Please take another look and give feedback.

MichaelBroughton commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,327 @@
+
+class TransformerStatsLogger:
+    """Abstract Base Class for transformer logging infrastructure.
+
```
Do we want to make a note here that this also provides the default terminal/print based logger behavior ?

MichaelBroughton commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,327 @@
+    def show(self, level: LogLevel = LogLevel.INFO) -> None:
+        """Show the stored logs >= level in the desired format."""
+
+        def print_log(log: LoggerNode, pad=''):
```
Nit: Can we move this out to a protected class level function with a leading underscore instead of a nested function ?

tanujkhattar replied:
I chose this to be a nested method because derived classes will be overriding the `show` method and then the inherited `_print` will become moot.

MichaelBroughton commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,327 @@
+            print(textwrap.indent(str(log.final_circuit), pad))
+            print("----------------------------------------")
+
+        done = [0] * self._curr_id
+
+        def dfs(log: LoggerNode, pad=''):
+            print_log(log, pad)
+            done[log.transformer_id] = True
+            for child_id in log.nested_loggers:
+                dfs(self._logs[child_id], pad + ' ' * 4)
+
+        for i in range(self._curr_id):
+            if not done[i]:
+                dfs(self._logs[i])
```
Can we swap this over to the iterative version ? I personally get pretty worried seeing recursion in library code unless it's absolutely needed.

MichaelBroughton commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,327 @@
+    """
+    if isinstance(cls_or_func, type):
+        return transformer_class(cls_or_func)
+    assert callable(
+        cls_or_func
+    ), f"The decorated object {cls_or_func} must be a callable class or a method."
```
Should this be at the top of the function body ?

MichaelBroughton commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,327 @@
+        self._logs: List[LoggerNode] = []
+        self._stack: List[int] = []
+
+    def register_initial(self, circuit: 'cirq.AbstractCircuit', transformer_name: str) -> None:
```
Should the circuit type here be abstractcircuit or CIRCUIT_TYPE to match register_final ?

tanujkhattar replied:
Since the logger will be shared across all transformers, we should use `cirq.AbstractCircuit` instead of `CIRCUIT_TYPE`. Updated `register_final` to match this.

MichaelBroughton commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,327 @@
+class TransformerContext:
+    """Stores common configurable options for transformers.
+    Args:
+        logger: `TransformerStatsLogger` instance, which is a stateful logger used for logging the
```
Nit: \`cirq.TransformerStatslogger\` for docs linking.

MichaelBroughton commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,327 @@
+    >>>    ...
+
+    Args:
+        func: Transformer implemented as a method from (CIRCUIT_TYPE, Context) --> CIRCUIT_TYPE.
```
Nit: docs linking for CIRCUIT_TYPE and Context here

MichaelBroughton commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,327 @@
+def transformer_method(func: TRANSFORMER_TYPE[CIRCUIT_TYPE]) -> TRANSFORMER_TYPE[CIRCUIT_TYPE]:
+    """Decorator to verify API and append logging functionality to callable transformer methods.
+
+    For example:
```
I think we need a space here in order for this to render correctly thought I'm not 100% sure.

MichaelBroughton commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,327 @@
+
+    The decorated class must implement the `__call__` method satisfying the `TRANSFORMER_TYPE` API.
+
+    For example:
```
Nit: space

MichaelBroughton commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,327 @@
+
+    Args:
+        cls: Transformer implemented as a class overriding the `__call__` method, which satisfies
+            the `TRANSFORMER_TYPE` API.
```
Nit: \`cirq.TRANSFORMER_TYPE\`

MichaelBroughton reviewed (commented):
This looks good to me after some nits. Quick question: Do we expect to need a lot of class based transformers with `__call__` defined ? Would we be able to keep things simple and just make them all standalone functions ?

maffoo commented on cirq-core/cirq/transformers/transformer_api_test.py:
```diff
@@ -0,0 +1,275 @@
+    if not hasattr(mock_transformer_method_circuit, 'mock'):
+        mock_transformer_method_circuit.mock = mock.Mock()  # type: ignore
+    mock_transformer_method_circuit.mock(circuit, context)  # type: ignore
+    return circuit
```
I'd suggest having a factory to create these transformers instead, so the mock is not part of a static object:
```python
def make_circuit_transformer_func() -> TRANSFORMER['cirq.Circuit']:
    my_mock = mock.Mock()
    @cirq.transformer
    def func(circuit: cirq.Circuit, context: cirq.TransformerContext) -> cirq.Circuit:
        my_mock(circuit, context)
        return circuit
    func.mock = my_mock
    return func
...
@pytest.mark.parametrize('transformer', [make_circuit_transformer_func(), ...])
def test...
```
This will avoid bugs if we add more tests that use circuit transformer funcs. Same thing for frozen circuit transformer funcs below.

maffoo commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,327 @@
+    return circuit
+
+
+def transformer_method(func: TRANSFORMER_TYPE[CIRCUIT_TYPE]) -> TRANSFORMER_TYPE[CIRCUIT_TYPE]:
```
The "method" terminology comes up in several places and should probably be changed to "function" or "func" throughout since this works for freestanding functions ("methods" are more specifically functions on classes).

Though I might suggest inlining the contents of this and `transformer_class` into `transformer` because I actually think having them close together makes more clear what's going on.

tanujkhattar replied:
inlined the contents

tanujkhattar commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,327 @@
+    initial_circuit: 'cirq.AbstractCircuit'
+    final_circuit: 'cirq.AbstractCircuit'
+    logs: Tuple[Tuple[LogLevel, Tuple[str, ...]], ...] = ()
+    nested_loggers: Tuple[int, ...] = ()
```
My original preference was because tuples are hashable / serializable but it doesn't really matter here so changed it to lists.

tanujkhattar commented:
@MichaelBroughton @maffoo I've made all changes as per the feedback. The major change now is the change in transformer API that I described in https://github.com/quantumlib/Cirq/pull/4797#discussion_r785184342

>Do we expect to need a lot of class based transformers with __call__ defined ? Would we be able to keep things simple and just make them all standalone functions?

@MichaelBroughton The major problem in keeping transformers as standalone functions right now is the fact that we don't support additional arguments except circuit and context due to limitations in mypy. 

It should be possible to add support for additional args in the decorator using `Concatenate` and `ParamSpec` introduced in https://www.python.org/dev/peps/pep-0612/.  The mypy bug which tracks the implementation is still open https://github.com/python/mypy/issues/8645

maffoo commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,287 @@
+    """Helper to log initial and final circuits before and after calling the transformer."""
+
+    context.logger.register_initial(circuit, transformer_name)
+    transformed_circuit = func(circuit.unfreeze(copy=True), context)
```
The design doc notes that "The callable should not mutate the input circuit and instead return a copy." With the current implementation if I chain two transformer functions, the first has to allocate and return a new circuit, but then it gets copied here before being passed into the second transformer, even though the second transformer is not supposed to mutate it. This copy seems like unnecessary work.

It seems to me that transformer functions should accept and return `cirq.AbstractCircuit` instead. Since this abstract base type does not have any mutation methods, that makes clear in the signature that properly-implemented transformers can't mutate the circuit they receive, and we don't do any unnecessary copying or converting between formats. I know it's _possible_ that we might introduce new circuit implementations in the future, and some transformers may need to convert to a specific format to be implemented most efficiently, rather than using only what is available in the `AbstractCircuit` interface, but it still seems to me preferable to make any such copies or format conversions explicit in the transformers that need them, rather than trying to bake any default conversions or copies into the logging machinery, as here.

tanujkhattar replied:
My intention to do this change was that with the current proposed design, the callable's are free to mutate the input as we have already enforced "don't mutate the input" part in the decorator by passing a copy to the actual transformer.

This has the following benefits:
1.  The contract is very precise -- there's no scope of questions like "The method receives and returns an `AbstractCircuit`, will it preserve the type I pass in? Can I back `FrozenCircuit` if I pass a `Circuit`? etc.
2.  Returning an `AbstractCircuit` has an another disadvantage that we lose the circuit type information that was used to call the particular transformer. And hence, we'll have to introduce explicit casts at the call site, which will be annoying for users. For example:
```
@cirq.transformer
def custom_transformer(circuit: 'cirq.AbstractCircuit') -> 'cirq.AbstractCircuit':
    pass

def user_helper_method(circuit: 'cirq.Circuit') -> 'cirq.Circuit':
    # Do something. 
    return custom_transformer(circuit) # This breaks and need's an explicit cast.
```

3. The pattern followed today is that optimizers mutate the input. If we continue to allow this and silently add a copy in the decorator, it makes the migration of all user defined transformers to be just "add a @cirq.transformer decorator and your are good". 



Given the above benefits, my proposal would be to 
1. Keep `TRANSFORMER_TO_DECORATE` and `TRANSFORMER` as it is. 
2. Maybe add an explicit flag to the `@cirq.transformer(copy_input=True)` which can be used to control whether we copy the input by default or not. The default will be set to `True` and if we write a transformer where this is not needed, we can set the flag to `False` by default. 


WDYT ?

maffoo reviewed (commented):
Thank you for simplifying the types here; it is much easier to understand what is going on. I do still have a concern about using `cirq.Circuit` by default, especially since the implementation is copying the circuit on each transformer invocation, despite the requirement that transformers not mutate their inputs. I'd suggest using `AbstractCircuit` which should remove the need for separate `TRANSFORMER` and `TRANSFORMER_TO_DECORATE` types. But I think this is very close.

dstrain115 commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,287 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
Optional:  Should this file be called api.py to avoid smurf naming?

tanujkhattar replied:
I think it's better to call it `transformer_api` because searching file named `api` in IDEs during development would be tricky. We are also using similar naming elsewhere - like `transformer_primitives` instead of just `primitives`.

dstrain115 commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,287 @@
+
+
+class LogLevel(enum.Enum):
+    """Different logging resolution options for `cirq.TransformerStatsLogger`.
```
This should still be clarified.  Does specifying DEBUG also include INFO and WARNING messages?

tanujkhattar replied:
This is specified below in the docstring of `TransformerLogger.show`, which is used to display the logs.

tanujkhattar replied:
Added another description here as well.

dstrain115 commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,287 @@
+
+
+@dataclasses.dataclass
+class LoggerNode:
```
Is this meant to be used externally?  Should we call it _LoggerNode?

tanujkhattar replied:
done

dstrain115 commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,287 @@
+
+@dataclasses.dataclass
+class LoggerNode:
+    """Stores logging data of a single transformer stage in `cirq.TransformerStatsLogger`."""
```
Add docstrings for arguments.

tanujkhattar replied:
done

tanujkhattar commented on cirq-core/cirq/transformers/transformer_api.py:
```diff
@@ -0,0 +1,287 @@
+    nested_loggers: List[int] = dataclasses.field(default_factory=list)
+
+
+class TransformerStatsLogger:
```
done

tanujkhattar commented:
@dstrain115 Addressed all your comments. 

@maffoo Incorporated your suggestions, as per our offline discussions,  for further simplifying the mypy types. A transformer is now just 

```python
TRANSFORMER = Callable[['cirq.AbstractCircuit', TransformerContext], 'cirq.AbstractCircuit']
```

and the decorator implementation ensures that we don't modify the input type of a callable to preserve stricter return types wherever possible. 

There is still a small glitch that the transformer type does not enforce any constraint on the argument names. This is only a problem in a small corner case where 

```python
@cirq.transformer
def f(c: cirq.AbstractCircuit, t: cirq.TransformerContext) -> cirq.Circuit:
    pass

f(circuit, context) # This works fine because positional arguments are used for invocation. 
f(c = circuit, t = context) # This will fail because we are explicitly using keyword arguments. 
```

To fix this, we'll have to change the transformer type to a protocol instead of a callable so we can enforce constraints on the argument names as well. However, I'd suggest let's do that in a separate PR as it's a small corner case and checking in this PR will help unblock all migration of all existing transformers. 

WDYT ?

maffoo reviewed (approved):
I think the internals of the `trasformer` decorator can be cleaned up a bit, but that doesn't affect the exported API, so we can address that in a follow-up. LGTM!

tanujkhattar commented:
Thanks @maffoo! Merging this now.
